src/main/java/org/terasology/web/resources/worldMap/WorldMapResource.java

```diff
@@ -0,0 +1,200 @@
+import org.terasology.web.resources.base.ResourceAccessException;
+import org.terasology.web.resources.base.ResourceMethod;
+import org.terasology.web.resources.base.ResourcePath;
+import org.terasology.world.RelevanceRegionComponent;
+import org.terasology.world.WorldProvider;
+import org.terasology.world.block.Block;
+import org.terasology.world.block.tiles.BlockTile;
+import org.terasology.world.chunks.ChunkConstants;
+import org.terasology.world.chunks.ChunkProvider;
+
+import javax.imageio.ImageIO;
+import java.awt.Color;
+import java.awt.Graphics2D;
+import java.awt.image.BufferedImage;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+
+import static org.terasology.web.resources.base.ResourceMethodFactory.createParameterlessMethod;
+
+public class WorldMapResource extends AbstractSimpleResource {
+
+    private static final Logger logger = LoggerFactory.getLogger(WorldMapResource.class);
+
+    @In
+    private WorldProvider worldProvider;
+
+    @In
+    private EntityManager entityManager;
+
+    // TODO: reimplement this as a GET request with parameters
+    @Override
+    protected ResourceMethod<WorldMapInput, String> getPutMethod(ResourcePath path) throws ResourceAccessException {
+        return createParameterlessMethod(path, ClientSecurityRequirements.PUBLIC, WorldMapInput.class,
+                (data, client) -> getWorldMapBase64ImageString(data.getCenter(), data.getMapBlockWidth(), data.getMapBlockLength(), data.isSurface(), client.getEntity()));
+    }
+
+    private String getWorldMapBase64ImageString(Vector3i center, int mapBlockWidth, int mapBlockLength, boolean isSurface, EntityRef clientEntity) {
+        final int colorSizeMultiplier = mapBlockWidth * mapBlockLength <= 125 * 125 ? 60 : 30;
+        int blockY = 40;
+        List<List<Color>> colors = new ArrayList<>(mapBlockWidth);
+        for (int i = 0; i < mapBlockWidth; ++i) {
+            colors.add(i, new ArrayList<>(mapBlockLength));
+        }
+
+
+        loadChunks(center, mapBlockWidth, mapBlockLength);
+
+        for (int x = (int) Math.floor((double) center.getX() - mapBlockWidth / 2); x < (int) Math.ceil((double) mapBlockWidth / 2 + center.getX()); ++x) {
+            for (int z = (int) Math.floor((double) center.getZ() - mapBlockLength / 2); z < (int) Math.ceil((double) mapBlockLength / 2 + center.getZ()); ++z) {
+                /*synchronized (this) {
+                    while (worldProvider.getBlock(x, blockY, z).getURI().toString().equals("engine:unloaded")) {
+                        try {
+                            wait(1000);
+                        } catch (InterruptedException e) {
+                            e.printStackTrace();
```

Comment:
We still need a way to give time for the chunks to load after chunk loading is triggered. You can see my first attempt here, which didn't have good enough performance.